### ISSUE TITLE
Silent incorrect result for addmm for noncontiguous input

### ISSUE THREAD

**Issue by tugsbayasgalan** [pytorch/pytorch#104439]
### 🐛 Describe the bug

```python
import torch

def f_non_contiguous(x, y, z):
    z_t = torch.ops.aten.t.default(z)
    return torch.ops.aten.addmm.default(x, y, z_t)

def f_contiguous(x, y, z):
    z_t = torch.ops.aten.t_copy.default(z)
    return torch.ops.aten.addmm.default(x, y, z_t)

x = torch.randn(256)
y = torch.randn(6, 960)
z = torch.randn(256, 960)

contiguous_result = f_contiguous(x, y, z)
non_contiguous_result = f_non_contiguous(x, y, z)

print(contiguous_result.is_contiguous())  # prints True 
print(non_contiguous_result.is_contiguous()) # print True 

print(torch.allclose(contiguous_result, non_contiguous_result)) # prints False
print(torch.allclose(contiguous_result.contiguous(), non_contiguous_result.contiguous())) # prints False
```

### Versions

latest main

cc @ezyang @gchanan @zou3519 @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @frank-wei

**Comment by janeyx99** [comment]
Can't repro on my current dev installation on commit 8a0873321889331348581625e91d9f26d68c51b0 from Monday 6/26 --> they all print True for me.

**Comment by janeyx99** [comment]
Upon @mikaylagawarecki's observation, we notice that a potential culprit is MKLDNN. My local build did not repro because I don't build with MKLDNN (aka `print(torch.backends.mkldnn.is_available())` is False) whereas @tugsbayasgalan's issue has `print(torch.backends.mkldnn.is_available())` returning True.

**Comment by mikaylagawarecki** [comment]
+1 able to reproduce the issue when building with USE_MKLDNN=1

**Comment by bdhirsh** [comment]
I tried shrinking the repro, and the smallest size can I repro successfully with is:
```
import torch

def f_non_contiguous(x, y, z):
    z_t = torch.ops.aten.t.default(z)
    return torch.ops.aten.addmm.default(x, y, z_t)

def f_contiguous(x, y, z):
    z_t = torch.ops.aten.t.default(z)
    return torch.ops.aten.addmm.default(x, y, z_t.contiguous())

x = torch.randn(2, 48)
y = torch.randn(2, 48)
z = torch.randn(48, 48)

contiguous_result = f_contiguous(x, y, z)
non_contiguous_result = f_non_contiguous(x, y, z)
print(torch.allclose(contiguous_result, non_contiguous_result)) # prints False
```

Running with sizes less than (2,48) seems to give correct results.

**Comment by mikaylagawarecki** [comment]
cc @jgong5 do you know who would be best able to help here?

**Comment by mingfeima** [comment]
@CaoE could you please help fix this one ?
if mkldnn can not handle this type of input at the first place, forbid it on such case.

**Comment by sanchitintel** [comment]
@mingfeima, I used Brian's script. oneMKL (not mkl-dnn) SGEMM kernels are being used in both cases.
The result is non-deterministic. eg. below, two consecutive runs of the script returned `True` & `False` respectively-
```
sanchitj@dev$ MKL_VERBOSE=1 python addmm.py
MKL_VERBOSE oneMKL 2022.0 Update 2 Product build 20220804 for Intel(R) 64 architecture Intel(R) Advanced Vector Extensions 512 (Intel(R) AVX-512) with support of Intel(R) Deep Learning Boost (Intel(R) DL Boost) and bfloat16, Lnx 3.30GHz lp64 gnu_thread
MKL_VERBOSE SGEMM(N,N,48,2,48,0x7ffcf5539e10,0x562884f79100,48,0x562884f6d480,48,0x7ffcf5539e30,0x562884f6db00,48) 49.15us CNR:OFF Dyn:1 FastMM:1 TID:0  NThr:104
MKL_VERBOSE SGEMM(T,N,48,2,48,0x7ffcf5539e10,0x562884f76c80,48,0x562884f6d480,48,0x7ffcf5539e30,0x562884f79100,48) 3.44us CNR:OFF Dyn:1 FastMM:1 TID:0  NThr:104
True
sanchitj@dev$ MKL_VERBOSE=1 python addmm.py
MKL_VERBOSE oneMKL 2022.0 Update 2 Product build 20220804 for Intel(R) 64 architecture Intel(R) Advanced Vector Extensions 512 (Intel(R) AVX-512) with support of Intel(R) Deep Learning Boost (Intel(R) DL Boost) and bfloat16, Lnx 3.30GHz lp64 gnu_thread
MKL_VERBOSE SGEMM(N,N,48,2,48,0x7fffc71c2ab0,0x55a388519280,48,0x55a38850d600,48,0x7fffc71c2ad0,0x55a38850dcc0,48) 70.62us CNR:OFF Dyn:1 FastMM:1 TID:0  NThr:104
MKL_VERBOSE SGEMM(T,N,48,2,48,0x7fffc71c2ab0,0x55a388516e00,48,0x55a38850d600,48,0x7fffc71c2ad0,0x55a388519280,48) 2.32us CNR:OFF Dyn:1 FastMM:1 TID:0  NThr:104
False
```

It seems that for that code-snippet, the result sometimes exceeds tolerance bounds.

However, the code snippet in the description of this GitHub issue is fully representative of the issue, in that it always produced consistent results at my end, and the deviation between both outputs (`contiguous_result` and `non_contiguous_result`) in the code is more than what the default tolerance allows.

**Comment by sanchitintel** [comment]
Hi @CaoE, please advise on a solution. Thanks!

**Comment by CaoE** [comment]
I have similar observations. The output with non-contiguous input sometimes exceeds tolerance bounds.

```
MKL_VERBOSE oneMKL 2022.0 Update 1 Product build 20220311 for Intel(R) 64 architecture Intel(R) Advanced Vector Extensions 512 (Intel(R) AVX-512) with Intel(R) DL Boost, bfloat16 support, Intel(R) AMX with bfloat16 and 8-bit integer support, and FP16 instructions, Lnx 1.80GHz lp64 gnu_thread
MKL_VERBOSE SGEMM(N,N,256,6,600,0x7ffd4203db40,0x5584d8260c80,256,0x5584d80fcc40,600,0x7ffd4203db60,0x5584d82f6d00,256) 7.03ms CNR:OFF Dyn:1 FastMM:1 TID:0  NThr:112
MKL_VERBOSE SGEMM(T,N,256,6,600,0x7ffd4203db40,0x5584d81cac00,600,0x5584d80fcc40,600,0x7ffd4203db60,0x5584d8260c80,256) 264.15us CNR:OFF Dyn:1 FastMM:1 TID:0  NThr:112

Mismatched elements: 13 / 1536 (0.8%)
Greatest absolute difference: 2.384185791015625e-05 at index (4, 203) (up to 1e-05 allowed)
Greatest relative difference: 0.0003173555014654994 at index (0, 99) (up to 1.3e-06 allowed)
```

**Comment by CaoE** [comment]
MKL SGEMM doesn't specify the computation order and it will have some accumulation errors by each implementation. The tolerance bounds seems strict for this case.

**Comment by tugsbayasgalan** [comment]
@CaoE Just wondering if this task is still being worked on?

**Comment by CaoE** [comment]
@tugsbayasgalan As I mentioned earlier, MKL will have some accumulation errors by each implementation. Do you need the results to be strictly equal for this case ？

**Comment by CaoE** [comment]
@tugsbayasgalan Can we close this issue ？The inconsistent result is an accuracy error, which is due to different implementations of MKL for contiguous inputs and noncontiguous inputs. Is such accuracy error acceptable in your case ? If it is not acceptable , you can use contiguous inputs instead.

**Comment by CaoE** [comment]
Close this issue and you can reopen it if you have any questions.
